### PR TITLE
make shell.nix better

### PR DESCRIPTION
**Pull request by Nilstrieb**
This makes it less opinionated.

cc @Nadrieril @fee1-dead i know you're using nixos

### CODE REVIEW

**Comment by 5225225** [comment]
One thing that would be useful is somehow getting `x` in the PATH and having it work.

Just doing 

```nix
x_application = pkgs.writeShellApplication {
  name = "x";
  text = ''exec ${./x} "$@"'';
};
```
and putting that in `nativeBuildInputs` doesn't work, since the script looks up stuff relative to itself, which is fair enough.

```nix
x_application = pkgs.writeShellApplication {
  name = "x";
  text = ''
      cd "${builtins.toString ./.}"
      ./x "$@"
  '';
};
```
seems to work fine? I don't know nix or bootstrap terribly well so there's probably a cleaner way of doing this.

**Comment by Nadrieril** [comment]
Oh RUSTC_ICE is the setting I didn't know I needed x) So many files turning up for no apparent reason!

I agree with most of the changes except that I really liked the RUSTC_BOOTSTRAP_CONFIG example. I wouldn't have known it was an option but now I can share my config.toml between machines. You could perhaps edit the comment instead, to have it say "this is an example config.toml, edit at your convenience"?

**Comment by rowan-sl** [comment]
In order to fix #1812 `nativeBuildInputs` can be changed to include `glibc.out glibc.static` (*in that order*)
 - additionally, `gcc_multi` can be removed

 for some reason `glibc.out` doesn't create the issues with `stdlib.h` that `glibc` did - [see this discussion](https://discourse.nixos.org/t/how-to-get-this-basic-c-build-to-work-in-a-nix-shell/12262/3)

**Comment by Nadrieril** [comment]
I added `gcc_multi` because it was required to run the `mir_opts` tests (see [this discussion](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Running.20mir-opt.20on.20nixos))

**Comment by eopb** [comment]
I generally agree with this change but I think it would be a shame to get rid of the recommendation of `patch-binaries-for-nix = true`.

That line is needed for nix shells on non-NixOS systems because [bootstrap doesn't have perfect nix detection](https://github.com/eopb/rust/blob/b8c207435c85955d0c0806240d5491f5ccd1def5/src/bootstrap/bootstrap.py#L651). I added it in https://github.com/rust-lang/rustc-dev-guide/pull/1774.

If we are going to get rid of the `config.toml` part of this `shell.nix` we should probably recommend `patch-binaries-for-nix` separately on the same page.

**Comment by fee1-dead** [comment]
Thanks for the ping. I actually don't use a shell.nix for hacking the rust compiler, since I thought it would be more hassle free if I installed these required things globally.

It might be helpful to showcase that using nix to manage config.toml is possible, but leave a comment saying that if anyone wants to change the behavior to use the `./config.toml` instead they can remove it.

**Comment by Nilstrieb** [comment]
> In order to fix #1812 `nativeBuildInputs` can be changed to include `glibc.out glibc.static` (_in that order_)
> 
>     * ~additionally, `gcc_multi` can be removed~ Looks like it was used in a different part than the part of the test suite i ran
> 
> 
> for some reason `glibc.out` doesn't create the issues with `stdlib.h` that `glibc` did - [see this discussion](https://discourse.nixos.org/t/how-to-get-this-basic-c-build-to-work-in-a-nix-shell/12262/3)

lol, i tried this out with awesome results: compiling rustc with this causes a rustc that just hangs (when cargo is invoking it to figure out cfgs etc). no problem, I try to `perf top` to figure out what's going on.

`perf: Segmentation fault`

**Comment by rowan-sl** [comment]
> > In order to fix #1812 `nativeBuildInputs` can be changed to include `glibc.out glibc.static` (_in that order_)
> > ```
> > * ~additionally, `gcc_multi` can be removed~ Looks like it was used in a different part than the part of the test suite i ran
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > for some reason `glibc.out` doesn't create the issues with `stdlib.h` that `glibc` did - [see this discussion](https://discourse.nixos.org/t/how-to-get-this-basic-c-build-to-work-in-a-nix-shell/12262/3)
> 
> lol, i tried this out with awesome results: compiling rustc with this causes a rustc that just hangs (when cargo is invoking it to figure out cfgs etc). no problem, I try to `perf top` to figure out what's going on.
> 
> `perf: Segmentation fault`

it has this problem when using the fix I gave? (or did you mean when `gcc_multi` is removed?)

**Comment by Nilstrieb** [comment]
ok my perf is always broken somehow (???). I did remove `gcc_multi` because it's useless now (as mir opt tests now don't require C anymore) but the change I made is adding `glibc.out glibc.static` to `nativeBuildInputs` (shouldn't it be `buildInputs` too?) which makes programs (build scripts, rustc) hang... will debug it further and report back
@Nemo157 mentioned that it works though

**Comment by Nilstrieb** [comment]
moving it to `buildInputs` fixed the hangs. i do not want to know where they come from. but i will put it there, that seems more appropriate anyways.

**Comment by Nadrieril** [comment]
Can you try running `./x.py test tests/mir-opts --bless` just to be sure? I remember `gcc_multi` being required for entirely silly reasons (bless is important because that runs more tests than without it 😅)

**Comment by Nilstrieb** [comment]
saethlin very recently made this not required anymore, being able to bless without a C toolchain.

**Comment by Nadrieril** [comment]
Ah I see, amazing then :D

File: src/building/suggested.md [hunk 1]
```diff
@@ -270,67 +270,34 @@ If you're using nix, you can use the following nix-shell to work on Rust:
...
-    in
-    pkgs.writeText "rustc-ripgreprc" "--ignore-file=${ignoreFile}";
-in
 pkgs.mkShell {
   name = "rustc";
   nativeBuildInputs = with pkgs; [
-    gcc_multi binutils cmake ninja openssl pkgconfig python39 git curl cacert patchelf nix psutils
+    binutils cmake ninja pkg-config python3 git curl cacert patchelf nix
+  ];
+  buildInputs = [
```
Comment (5225225) [inline src/building/suggested.md:278]: this needs `with pkgs;`

**Comment by Nilstrieb** [comment]
This should be fine to merge now. We could also mention this in `x setup` when it detects nix (which it already) can, but that's something for later.

**Review by Nadrieril — APPROVED** [review]

**Pull request merged**